synapse/http/proxyagent.py

```diff
@@ -194,24 +203,21 @@ def _http_proxy_endpoint(proxy, reactor, **kwargs):
     if proxy is None:
         return None
 
-    # currently we only support hostname:port. Some apps also support
-    # protocol://<host>[:port], which allows a way of requiring a TLS connection to the
-    # proxy.
+    parsed_url = urlsplit(proxy, scheme=b"http")
+    scheme = parsed_url.scheme
+    host = parsed_url.hostname
+    port = parsed_url.port
 
-    host, port = parse_host_port(proxy, default_port=1080)
-    return HostnameEndpoint(reactor, host, port, **kwargs)
+    if scheme not in (b"http", b"https"):
+        raise ValueError("Proxy scheme '%s' not supported", scheme)
```

Comment:
I am not sure if raising a ValueError is in any way appropriate here. Please advise :).

Reply:
It is probably reasonable, but we would want to:

1. Document this in the docstring.
2. Check that callers do a reasonable thing when trying to instantiate this.

Usually I would suggesting raising a `ConfigError` somewhere, but proxies aren't set via the config so I don't think that would be appropriate?

Reply:
I documented this and checked what happens when actually trying to do this in the real-world.

In my test this brought down synapse during startup showing this:

```
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,518 - twisted - 254 - ERROR - None - Traceback (most recent call last):
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,519 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,519 - twisted - 254 - ERROR - None -     return _run_code(code, main_globals, None,
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,519 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,519 - twisted - 254 - ERROR - None -     exec(code, run_globals)
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,519 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/app/homeserver.py", line 507, in <module>
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,520 - twisted - 254 - ERROR - None -     main()
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,520 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/app/homeserver.py", line 502, in main
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,520 - twisted - 254 - ERROR - None -     hs = setup(sys.argv[1:])
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,520 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/app/homeserver.py", line 360, in setup
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,520 - twisted - 254 - ERROR - None -     hs.setup()
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,521 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 273, in setup
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,521 - twisted - 254 - ERROR - None -     self.setup_background_tasks()
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,521 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 282, in setup_background_tasks
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,521 - twisted - 254 - ERROR - None -     getattr(self, "get_" + i + "_handler")()
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,521 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 172, in _get
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,522 - twisted - 254 - ERROR - None -     dep = builder(self)
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,522 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 438, in get_auth_handler
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,522 - twisted - 254 - ERROR - None -     return AuthHandler(self)
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,522 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/handlers/auth.py", line 179, in __init__
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,522 - twisted - 254 - ERROR - None -     inst = auth_checker_class(hs)
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/handlers/ui_auth/checkers.py", line 84, in __init__
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -     self._http_client = hs.get_proxied_http_client()
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 172, in _get
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -     dep = builder(self)
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/server.py", line 363, in get_proxied_http_client
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,523 - twisted - 254 - ERROR - None -     return SimpleHttpClient(
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,524 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/http/client.py", line 342, in __init__
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,524 - twisted - 254 - ERROR - None -     self.agent = ProxyAgent(
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,524 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/http/proxyagent.py", line 89, in __init__
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,524 - twisted - 254 - ERROR - None -     self.http_proxy_endpoint = _http_proxy_endpoint(
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,524 - twisted - 254 - ERROR - None -   File "/usr/lib/python3.9/site-packages/synapse/http/proxyagent.py", line 212, in _http_proxy_endpoint
Feb 01 22:17:56 matrix synapse[9335]: 2021-02-01 22:17:56,525 - twisted - 254 - ERROR - None -     raise ValueError("Proxy scheme 'socks not supported")
```
While maybe not being the nicest error message it is at least immediately obvious what is wrong when trying this.